src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp

```diff
@@ -357,66 +359,62 @@ namespace
                     abort_switch,
                     batch_size,
                     framebuffer,
-                    second_framebuffer,
+                    second_framebuffer.get(),
                     tile_origin_x,
                     tile_origin_y,
                     frame,
                     frame_width,
                     pass_hash,
                     aov_count);
 
-                // Check if this was the last batch.
-                if (remaining_samples - batch_size < 1)
+                const AABB2u block_image_bb = AABB2i::intersect(framebuffer->get_crop_window(), pb.m_surface);
+
+                // Evaluate block's noise amount.
+                pb.m_block_error = FilteredTile::compute_tile_variance(
+                    block_image_bb,
+                    framebuffer,
+                    second_framebuffer.get());
+
+                if (batch_size < m_params.m_batch_size)
                 {
+                    // It was the last batch.
                     finished_blocks.push_back(pb);
                 }
-                else
+                else if (pb.m_block_error <= m_params.m_noise_threshold)
                 {
-                    const AABB2u& block_image_bb = AABB2i::intersect(framebuffer->get_crop_window(), pb.m_surface);
-
-                    // Evaluate block's noise amount.
-                    pb.m_block_error = FilteredTile::compute_tile_variance(
-                        block_image_bb,
-                        framebuffer,
-                        second_framebuffer);
-
-                    // Decide if the blocks needs to be splitted, sampled or if it has converged.
-                    if (pb.m_block_error <= m_params.m_noise_threshold)
+                    // The block has converged.
+                    pb.m_converged = true;
+                    finished_blocks.push_back(pb);
+                }
+                else if (pb.m_block_error <= m_params.m_splitting_threshold)
+                {
+                    // The block needs to be splitted.
+                    if (pb.m_main_axis == PixelBlock::Axis::Horizontal
+                        && block_image_bb.extent(0) >= BlockSplittingThreshold)
                     {
-                        pb.m_converged = true;
-                        finished_blocks.push_back(pb);
+                        split_pixel_block(
+                            pb,
+                            rendering_blocks,
+                            static_cast<int>((block_image_bb.min.x + block_image_bb.max.x) * 0.5f));
```

Comment:
I believe that

    static_cast<int>((block_image_bb.min.y + block_image_bb.max.y) * 0.5f)

is equivalent to

    (block_image_bb.min.y + block_image_bb.max.y) / 2